app/scripts/controllers/transactions/IncomingTransactionHelper.ts

```diff
@@ -0,0 +1,271 @@
+import EventEmitter from 'events';
+import type { BlockTracker, NetworkState } from '@metamask/network-controller';
+import type { Hex } from '@metamask/utils';
+
+import log from 'loglevel';
+import { TransactionMeta } from '../../../../shared/constants/transaction';
+import { RemoteTransactionSource } from './types';
+
+const UPDATE_CHECKS: ((txMeta: TransactionMeta) => any)[] = [
+  (txMeta) => txMeta.status,
+];
+
+export class IncomingTransactionHelper {
+  hub: EventEmitter;
+
+  #blockTracker: BlockTracker;
+
+  #getCurrentAccount: () => string;
+
+  #getLocalTransactions: () => TransactionMeta[];
+
+  #getNetworkState: () => NetworkState;
+
+  #isEnabled: () => boolean;
+
+  #isRunning: boolean;
+
+  #lastFetchedBlockNumbers: Record<string, number>;
+
+  #onLatestBlock: (blockNumberHex: Hex) => Promise<void>;
+
+  #remoteTransactionSource: RemoteTransactionSource;
+
+  #transactionLimit?: number;
+
+  #updateTransactions: boolean;
+
+  constructor({
+    blockTracker,
+    getCurrentAccount,
+    getLocalTransactions,
+    getNetworkState,
+    isEnabled,
+    lastFetchedBlockNumbers,
+    remoteTransactionSource,
+    transactionLimit,
+    updateTransactions,
+  }: {
+    blockTracker: BlockTracker;
+    getCurrentAccount: () => string;
+    getNetworkState: () => NetworkState;
+    getLocalTransactions?: () => TransactionMeta[];
+    isEnabled?: () => boolean;
+    lastFetchedBlockNumbers?: Record<string, number>;
+    remoteTransactionSource: RemoteTransactionSource;
+    transactionLimit?: number;
+    updateTransactions?: boolean;
+  }) {
+    this.hub = new EventEmitter();
+
+    this.#blockTracker = blockTracker;
+    this.#getCurrentAccount = getCurrentAccount;
+    this.#getLocalTransactions = getLocalTransactions || (() => []);
+    this.#getNetworkState = getNetworkState;
+    this.#isEnabled = isEnabled ?? (() => true);
+    this.#isRunning = false;
+    this.#lastFetchedBlockNumbers = lastFetchedBlockNumbers ?? {};
+    this.#remoteTransactionSource = remoteTransactionSource;
+    this.#transactionLimit = transactionLimit;
+    this.#updateTransactions = updateTransactions ?? false;
+
+    // Using a property instead of a method to provide a listener reference
+    // with the correct scope that we can remove later if stopped.
+    this.#onLatestBlock = async (blockNumberHex: Hex) => {
+      try {
+        await this.#update(blockNumberHex);
+      } catch (error) {
+        log.error('Error while checking incoming transactions', error);
+      }
+    };
+  }
+
+  start() {
+    if (this.#isRunning) {
+      return;
+    }
+
+    if (!this.#canStart()) {
+      return;
+    }
+
+    this.#blockTracker.addListener('latest', this.#onLatestBlock);
+    this.#isRunning = true;
+  }
+
+  stop() {
+    this.#blockTracker.removeListener('latest', this.#onLatestBlock);
+    this.#isRunning = false;
+  }
+
+  async #update(latestBlockNumberHex?: Hex): Promise<void> {
+    if (!this.#canStart()) {
```

Comment:
Is there any situation that requires this check to be in "#update" since it's been checked on "start"? Or just an extra security?

Reply:
Once the polling is started via the `start` method, this method will be ran each time there is a new block so we still have to verify that incoming transactions are still enabled and that the current network is still supported in case they have been changed in the meantime.

The polling is also stopped and started when the preferences controller detects this has changed, but the controller (which will eventually be in core) can't assume the client is checking that.